### ISSUE TITLE
Add flexbox variant to Carousel

### ISSUE THREAD

**Issue by mdo** [twbs/bootstrap#21305]
After merging #21298, we're in a much better spot to create the flex variation for `$enable-flex`. Help appreciated if anyone wants to try to tackle this! :D

**Comment by aRestless** [comment]
Some food for thought: If in any way possible, it would be amazing if a flex version would allow for a constant height carousel, i.e. one that aligns its overall height with the highest contained element.

This would not only be useful to handle images with different aspect ratios, but would also allow to easily build a traditional wizard widget using the bootstrap carousels (which is my use-case currently).

I'm currently fiddling around with that, I'll let you know if I get any results.

**Comment by aRestless** [comment]
Alright, I got a working proof-of-concept which might be a good starting point for the actual implementation: https://jsbin.com/cupopobabe/edit?html,css,js,output

I must admit that I haven't looked too much at the current code for carousel yet, so I don't know how compliant this is, but it works without modifying the current HTML structure. Basically: All carousel-items are "displayed" the whole time, thereby affecting the height of the container, with a `flex` propery of `0 0 100%`, spanning the whole carousel width. Which item is actually visible is determined by the `order` property. The active item always has an `order` of 0, and while transitioning to the left or right those target items have an `order` of -1 and 1, respectively. All other items are given an absurdly high `order` value (like 10000), so they are inserted at the end of the container and don't shift around the relevant carousel-items.

For the actual animation I'm using a pseudo element at `order: -2` with `margin-left: -100%` (so that it doesn't push the active carousel-item out of the viewable space) and a transition of its `flex-basis`.

The result is that the `.carousel` always has the same height, and the individual carousel-items can be aligned arbitrarily - I used `align-items: center` in the example for demonstration purposes.

If having the constant height is not desired, `.carousel-item` can just default to `display: none` just as it is now and the behaviour is equivalent to the current non-flex version.
